File: cli/Sources/TuistServer/Models/Preview.swift
import Foundation
import TuistSimulator
import XcodeGraph

public struct Preview: Equatable, Codable, Identifiable {
    public let id: String
    public let url: URL
    public let qrCodeURL: URL
    public let iconURL: URL
    public let bundleIdentifier: String?
    public var displayName: String?
    public let gitCommitSHA: String?
    public let gitBranch: String?
    public let appBuilds: [AppBuild]
    public let supportedPlatforms: [DestinationType]
}

extension Preview {
    init?(_ preview: Components.Schemas.Preview) {
        id = preview.id
        guard let url = URL(string: preview.url),
              let qrCodeURL = URL(string: preview.qr_code_url),
              let iconURL = URL(string: preview.icon_url)
        else { return nil }
        self.url = url
        self.qrCodeURL = qrCodeURL
        self.iconURL = iconURL
        bundleIdentifier = preview.bundle_identifier
        displayName = preview.display_name
        gitCommitSHA = preview.git_commit_sha
        gitBranch = preview.git_branch
        appBuilds = preview.builds.compactMap(AppBuild.init)
        supportedPlatforms = preview.supported_platforms.map(DestinationType.init)
    }
}

#if DEBUG
    extension Preview {
        public static func test(
            id: String = "preview-id",
            url: URL = URL(string: "https://tuist.dev/tuist/tuist/previews/preview-id")!,
            // swiftlint:disable:this force_try,
            qrCodeURL: URL =
                URL(string: "https://tuist.dev/tuist/tuist/previews/preview-id/qr-code.svg")!,
            // swiftlint:disable:this force_try
            iconURL: URL =
                URL(string: "https://tuist.dev/tuist/tuist/previews/preview-id/icon.png")!,
            // swiftlint:disable:this force_try
            bundleIdentifier: String? = "dev.tuist.app",
            displayName: String? = "App",
            gitCommitSHA: String? = nil,
            gitBranch: String? = nil,
            appBuilds: [AppBuild] = [
                .test(),
            ],
            supportedPlatforms: [DestinationType] = [
                .device(.iOS),
                .simulator(.iOS),
            ]
        ) -> Self {
            .init(
                id: id,
                url: url,
                qrCodeURL: qrCodeURL,
                iconURL: iconURL,
                bundleIdentifier: bundleIdentifier,
                displayName: displayName,
                gitCommitSHA: gitCommitSHA,
                gitBranch: gitBranch,
                appBuilds: appBuilds,
                supportedPlatforms: supportedPlatforms
            )
        }
    }
#endif

File: server/lib/tuist_web/api/schemas/preview.ex
defmodule TuistWeb.API.Schemas.Preview do
  @moduledoc """
  The schema for the Tuist Preview response.
  """
  alias OpenApiSpex.Schema
  alias TuistWeb.API.Schemas.Account

  require OpenApiSpex

  OpenApiSpex.schema(%{
    type: :object,
    required: [
      :id,
      :url,
      :qr_code_url,
      :icon_url,
      :builds,
      :supported_platforms,
      :inserted_at,
      :created_from_ci,
      :device_url
    ],
    properties: %{
      id: %Schema{
        type: :string,
        description: "Unique identifier of the preview."
      },
      url: %Schema{type: :string, description: "The URL to download the preview"},
      device_url: %Schema{type: :string, description: "The URL to download the preview on a device"},
      qr_code_url: %Schema{
        type: :string,
        description: "The URL for the QR code image to dowload the preview"
      },
      icon_url: %Schema{
        type: :string,
        description: "The URL for the icon image of the preview"
      },
      version: %Schema{
        type: :string,
        description: "The app version of the preview"
      },
      bundle_identifier: %Schema{
        type: :string,
        description: "The bundle identifier of the preview"
      },
      display_name: %Schema{
        type: :string,
        description: "The display name of the preview"
      },
      git_commit_sha: %Schema{
        type: :string,
        description: "The git commit SHA associated with the preview"
      },
      git_branch: %Schema{
        type: :string,
        description: "The git branch associated with the preview"
      },
      track: %Schema{
        type: :string,
        description: "The track for the preview (e.g., 'beta', 'nightly')"
      },
      builds: %Schema{
        type: :array,
        items: TuistWeb.API.Schemas.AppBuild
      },
      supported_platforms: %Schema{
        type: :array,
        items: TuistWeb.API.Schemas.PreviewSupportedPlatform
      },
      inserted_at: %Schema{
        type: :string,
        format: :date_time,
        description: "The date and time when the preview was inserted"
      },
      created_by: Account,
      created_from_ci: %Schema{
        type: :boolean,
        description: "Whether the preview was created from CI"
      }
    }
  })
end
